File: src/main/java/org/cubexmc/railway/service/strategy/LocalDispatchStrategy.java
package org.cubexmc.railway.service.strategy;

import java.util.List;

import org.cubexmc.railway.Railway;
import org.cubexmc.railway.model.Line;
import org.cubexmc.railway.model.Stop;
import org.cubexmc.railway.service.DispatchStrategy;
import org.cubexmc.railway.service.LineService;
import org.cubexmc.railway.service.virtual.VirtualTrain;
import org.cubexmc.railway.service.virtual.VirtualTrainPool;
import org.cubexmc.railway.train.TrainInstance;

/**
 * Local mode dispatch: maintain a virtual railway network where trains run
 * continuously in a virtual layer. When a virtual train arrives at a station
 * with player demand, it is materialized into a physical train.
 * 
 * Key design: Virtual trains simulate real-time positions. Physical trains
 * only spawn when a virtual train ACTUALLY arrives at a station where
 * players are waiting.
 */
public class LocalDispatchStrategy implements DispatchStrategy {

    private VirtualTrainPool pool;
    private boolean initialized = false;

    // Track which stop has player demand (for materialization check)
    private String currentDemandStopId = null;

    // Cooldown to prevent rapid re-spawn
    private long lastSpawnTick = -1L;
    private static final long SPAWN_COOLDOWN_TICKS = 60; // 3 seconds

    @Override
    public void tick(LineService service, long currentTick) {
        // Initialize pool on first tick
        if (!initialized) {
            initializePool(service, currentTick);
            initialized = true;
        }

        if (pool == null) {
            return;
        }

        Railway plugin = service.getPlugin();
        Line line = service.getLine();
        if (line == null)
            return;
        List<String> stops = line.getOrderedStopIds();

        // First, clean up stale materialized markers
        cleanupStaleMaterializations(service);

        // Update player demand
        Stop demandStop = findPlayerOccupiedStop(service);
        currentDemandStopId = (demandStop != null) ? demandStop.getId() : null;

        // Check if we already have a physical train for this line
        boolean hasPhysicalTrain = !service.getActiveTrains().isEmpty();

        // Check spawn cooldown
        boolean onCooldown = lastSpawnTick > 0 && (currentTick - lastSpawnTick) < SPAWN_COOLDOWN_TICKS;

        // Update virtual train positions (this moves them forward in time)
        // IMPORTANT: Update pool BEFORE checking for materialization
        // This ensures that if a train was just cleaned up (and dwell expired),
        // it departs before we try to spawn it again.
        pool.tick(currentTick, plugin.getTravelTimeEstimator());

        // Try to materialize a virtual train that has ARRIVED at the demand stop
        if (currentDemandStopId != null && !hasPhysicalTrain && !onCooldown) {
            tryMaterializeArrivedTrain(service, currentTick, stops);
        }
    }

    /**
     * Try to materialize a virtual train that has arrived at the demand stop.
     * Only spawn if a virtual train is at (or very close to) the demand stop.
     */
    private void tryMaterializeArrivedTrain(LineService service, long currentTick, List<String> stops) {
        Railway plugin = service.getPlugin();
        int demandIndex = stops.indexOf(currentDemandStopId);
        if (demandIndex < 0)
            return;

        // Look for a virtual train that has arrived at the demand stop
        for (VirtualTrain vt : pool.getVirtualTrains()) {
            if (pool.isMaterialized(vt.getId())) {
                continue;
            }

            // Check if this virtual train is at the demand stop
            // It must be in WAITING state at the exact stop index
            if (vt.getState() == VirtualTrain.State.WAITING && vt.getCurrentStopIndex() == demandIndex) {
                // Virtual train has arrived! Materialize it.
                plugin.getLogger().info("[LocalDispatch] Virtual train " +
                        vt.getId().toString().substring(0, 8) + " arrived at " + currentDemandStopId);

                if (materializeTrain(service, vt, currentDemandStopId, demandIndex, stops, currentTick)) {
                    lastSpawnTick = currentTick;
                    return; // Only spawn one train
                }
            }
        }

        // No train has arrived yet - this is normal, player waits for ETA
        // Log occasionally to show system is working
        if (currentTick % 100 == 0) { // Every 5 seconds
            VirtualTrain closest = findClosestVirtualTrain(demandIndex, plugin);
            if (closest != null) {
                double eta = closest.estimateEtaToStop(demandIndex, plugin.getTravelTimeEstimator());
                plugin.getLogger().fine("[LocalDispatch] Waiting for train to arrive at " +
                        currentDemandStopId + ", ETA=" + String.format("%.1f", eta) + "s");
            }
        }
    }

    private VirtualTrain findClosestVirtualTrain(int demandIndex, Railway plugin) {
        VirtualTrain closest = null;
        double bestEta = Double.POSITIVE_INFINITY;

        for (VirtualTrain vt : pool.getVirtualTrains()) {
            if (pool.isMaterialized(vt.getId()))
                continue;
            double eta = vt.estimateEtaToStop(demandIndex, plugin.getTravelTimeEstimator());
            if (Double.isFinite(eta) && eta < bestEta) {
                bestEta = eta;
                closest = vt;
            }
        }
        return closest;
    }

    /**
     * Clean up materialized markers for virtual trains whose physical trains no
     * longer exist.
     */
    private void cleanupStaleMaterializations(LineService service) {
        if (pool == null)
            return;

        List<TrainInstance> activeTrains = service.getActiveTrains();
        List<VirtualTrain> virtualTrains = pool.getVirtualTrains();

        for (VirtualTrain vt : virtualTrains) {
            if (!pool.isMaterialized(vt.getId())) {
                continue;
            }

            // Check if any active train has this virtual train ID
            boolean found = false;
            for (TrainInstance ti : activeTrains) {
                if (vt.getId().equals(ti.getVirtualTrainId())) {
                    found = true;
                    break;
                }
            }

            // If no physical train found, un-materialize
            if (!found) {
                // Just clear the materialized flag - virtual train continues from its current
                // position
                pool.clearMaterialized(vt.getId());
                service.getPlugin().getLogger().info(
                        "[LocalDispatch] Cleared materialization for " + vt.getId().toString().substring(0, 8));
            }
        }
    }

    private void initializePool(LineService service, long currentTick) {
        Line line = service.getLine();
        if (line == null) {
            return;
        }

        pool = new VirtualTrainPool(service.getLineId(), service.getDwellTicks());
        pool.initialize(line, service.getHeadwaySeconds(),
                service.getPlugin().getTravelTimeEstimator(), currentTick);

        service.getPlugin().getLogger().info(
                "[LocalDispatch] Initialized pool for line " + service.getLineId() +
                        " with " + pool.getActiveCount(line.getOrderedStopIds()) + " virtual trains");
    }

    private Stop findPlayerOccupiedStop(LineService service) {
        Line line = service.getLine();
        if (line == null)
            return null;

        List<String> stopIds = line.getOrderedStopIds();
        Railway plugin = service.getPlugin();
        double radius = plugin.getLocalActivationRadius();
        double radiusSq = radius * radius;

        for (String sid : stopIds) {
            Stop s = plugin.getStopManager().getStop(sid);
            if (s == null || s.getStopPointLocation() == null)
                continue;
            if (plugin.isPlayerWithinStopRadius(s, radiusSq)) {
                return s;
            }
        }
        return null;
    }

    private boolean materializeTrain(LineService service, VirtualTrain vt,
            String targetStopId, int targetIndex, List<String> stops, long currentTick) {

        Railway plugin = service.getPlugin();

        // Don't spawn at terminal for non-loop lines
        if (targetIndex >= stops.size() - 1 && !service.isLoopLine()) {
            plugin.getLogger().fine("[LocalDispatch] Skipping terminal stop " + targetStopId);
            return false;
        }

        // Virtual train is at targetStop, spawn the physical train there
        int fromStopIndex = vt.getCurrentStopIndex();
        int toStopIndex = fromStopIndex; // Target the station we are at!

        // Spawn the physical train at the current stop
        TrainInstance train = service.spawnTrainForVirtual(
                currentTick, fromStopIndex, toStopIndex, 0.0,
                vt.getId(), targetStopId);

        if (train != null) {
            pool.markMaterialized(vt.getId());
            plugin.getLogger().info("[LocalDispatch] Train materialized at " + targetStopId);
            return true;
        } else {
            plugin.getLogger().warning("[LocalDispatch] Failed to spawn train at " + targetStopId);
            return false;
        }
    }

    /**
     * Return a physical train back to the virtual pool.
     */
    public void returnTrainToPool(TrainInstance train, long currentTick) {
        if (pool == null)
            return;

        java.util.UUID vtId = train.getVirtualTrainId();
        if (vtId == null)
            return;

        Line line = train.getService().getLine();
        if (line == null)
            return;
        List<String> stopIds = line.getOrderedStopIds();

        TrainInstance.VirtualizationState state = train.getVirtualizationState(currentTick);
        pool.returnToVirtual(vtId, state.currentIndex, state.targetIndex,
                state.progress, state.isWaiting, currentTick, stopIds);
    }

    public VirtualTrainPool getPool() {
        return pool;
    }

    @Override
    public void refreshTopology(LineService service, List<String> newStopIds, long currentTick) {
        if (pool == null)
            return;

        Railway plugin = service.getPlugin();
        pool.refreshTopology(newStopIds, plugin.getTravelTimeEstimator(), currentTick);
        plugin.getLogger().info("[LocalDispatch] Refreshed topology: " + newStopIds.size() + " stops");
    }
}

File: src/main/java/org/cubexmc/metro/service/strategy/LocalDispatchStrategy.kt
package org.cubexmc.metro.service.strategy

import org.cubexmc.metro.Metro
import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.service.DispatchStrategy
import org.cubexmc.metro.service.LineService
import org.cubexmc.metro.service.virtual.VirtualTrain
import org.cubexmc.metro.service.virtual.VirtualTrainPool
import org.cubexmc.metro.train.TrainInstance
import org.cubexmc.metro.util.SchedulerUtil
import java.util.UUID
import kotlin.math.max

/** Local mode keeps trains virtual until one reaches a stop with player demand. */
class LocalDispatchStrategy : DispatchStrategy {
    var pool: VirtualTrainPool? = null
        private set

    private var initialized = false
    private var currentDemandStopId: String? = null
    private val stopDemandExpiryTicks: MutableMap<String, Long> = HashMap()
    private var lastSpawnTick = -1L

    override fun tick(service: LineService, currentTick: Long) {
        if (!initialized) {
            initializePool(service, currentTick)
            initialized = true
        }
        val currentPool = pool ?: return
        val plugin = service.plugin
        val line = service.line ?: return
        val stops = line.orderedStopIds

        cleanupStaleMaterializations(service)
        currentDemandStopId = resolveDemandStop(service, currentTick, stops)
        val hasPhysicalTrain = service.activeTrains.isNotEmpty()
        val onCooldown = lastSpawnTick > 0 && currentTick - lastSpawnTick < 60L

        currentPool.tick(currentTick, plugin.travelTimeEstimator)
        if (currentDemandStopId != null && !hasPhysicalTrain && !onCooldown) {
            tryMaterializeArrivedTrain(service, currentTick, stops)
        }
    }

    override fun requestStop(service: LineService, stopId: String?, currentTick: Long) {
        val line = service.line
        if (line == null || stopId == null) {
            return
        }
        val stops = line.orderedStopIds
        val stopIndex = stops.indexOf(stopId)
        if (stopIndex < 0 || isTerminalDemand(service, stopIndex, stops)) {
            return
        }
        val keepAliveTicks = max(20L * 600L, service.headwayTicks * 2L)
        stopDemandExpiryTicks[stopId] = currentTick + keepAliveTicks
        currentDemandStopId = stopId
    }

    private fun resolveDemandStop(
        service: LineService,
        currentTick: Long,
        stops: List<String>,
    ): String? {
        cleanupExpiredDemands(currentTick)
        findPlayerOccupiedStop(service)?.let { occupiedStop ->
            requestStop(service, occupiedStop.id, currentTick)
        }
        if (stopDemandExpiryTicks.isEmpty()) {
            return null
        }

        var bestStopId: String? = null
        var bestEta = Double.POSITIVE_INFINITY
        val plugin = service.plugin
        for (stopId in stopDemandExpiryTicks.keys) {
            val stopIndex = stops.indexOf(stopId)
            if (stopIndex < 0 || isTerminalDemand(service, stopIndex, stops)) {
                continue
            }
            val candidate = findClosestVirtualTrain(stopIndex, plugin, currentTick)
            val eta = candidate?.estimateEtaToStop(
                stopIndex,
                plugin.travelTimeEstimator,
                currentTick,
            ) ?: Double.POSITIVE_INFINITY
            if (eta < bestEta) {
                bestEta = eta
                bestStopId = stopId
            }
        }
        return bestStopId
    }

    private fun cleanupExpiredDemands(currentTick: Long) {
        val iterator = stopDemandExpiryTicks.entries.iterator()
        while (iterator.hasNext()) {
            if (iterator.next().value <= currentTick) {
                iterator.remove()
            }
        }
    }

    private fun tryMaterializeArrivedTrain(
        service: LineService,
        currentTick: Long,
        stops: List<String>,
    ) {
        val demandStopId = currentDemandStopId ?: return
        val demandIndex = stops.indexOf(demandStopId)
        if (demandIndex < 0) {
            return
        }
        if (isTerminalDemand(service, demandIndex, stops)) {
            stopDemandExpiryTicks.remove(demandStopId)
            currentDemandStopId = null
            return
        }

        val currentPool = pool ?: return
        for (virtualTrain in currentPool.getVirtualTrains()) {
            if (currentPool.isMaterialized(virtualTrain.id)) {
                continue
            }
            if (virtualTrain.state == VirtualTrain.State.WAITING &&
                virtualTrain.currentStopIndex == demandIndex
            ) {
                if (materializeTrain(service, virtualTrain, demandStopId, demandIndex, stops, currentTick)) {
                    stopDemandExpiryTicks.remove(demandStopId)
                    lastSpawnTick = currentTick
                    return
                }
                lastSpawnTick = currentTick
                return
            }
        }

        if (currentTick % 100 == 0L) {
            val plugin = service.plugin
            val closest = findClosestVirtualTrain(demandIndex, plugin, currentTick)
            if (closest != null) {
                val eta = closest.estimateEtaToStop(
                    demandIndex,
                    plugin.travelTimeEstimator,
                    SchedulerUtil.getCurrentTick(),
                )
                plugin.logger.fine(
                    "[LocalDispatch] Waiting for train to arrive at $demandStopId, " +
                        "ETA=${String.format("%.1f", eta)}s",
                )
            }
        }
    }

    private fun findClosestVirtualTrain(
        demandIndex: Int,
        plugin: Metro,
        currentTick: Long,
    ): VirtualTrain? {
        val currentPool = pool ?: return null
        var closest: VirtualTrain? = null
        var bestEta = Double.POSITIVE_INFINITY
        for (virtualTrain in currentPool.getVirtualTrains()) {
            if (currentPool.isMaterialized(virtualTrain.id)) {
                continue
            }
            val eta = virtualTrain.estimateEtaToStop(
                demandIndex,
                plugin.travelTimeEstimator,
                currentTick,
            )
            if (eta.isFinite() && eta < bestEta) {
                bestEta = eta
                closest = virtualTrain
            }
        }
        return closest
    }

    private fun cleanupStaleMaterializations(service: LineService) {
        val currentPool = pool ?: return
        val activeTrains = service.activeTrains
        for (virtualTrain in currentPool.getVirtualTrains()) {
            if (!currentPool.isMaterialized(virtualTrain.id)) {
                continue
            }
            val found = activeTrains.any { train -> virtualTrain.id == train.virtualTrainId }
            if (!found) {
                currentPool.clearMaterialized(virtualTrain.id)
            }
        }
    }

    private fun initializePool(service: LineService, currentTick: Long) {
        val line = service.line ?: return
        pool = VirtualTrainPool(service.lineId, service.dwellTicks).also { newPool ->
            newPool.initialize(
                line,
                service.headwaySeconds,
                service.plugin.travelTimeEstimator,
                currentTick,
            )
        }
    }

    private fun findPlayerOccupiedStop(service: LineService): Stop? {
        val line = service.line ?: return null
        val plugin = service.plugin
        val radius = plugin.localActivationRadius
        val radiusSq = radius * radius
        for (stopId in line.orderedStopIds) {
            val stop = plugin.stopManager.getStop(stopId)
            if (stop?.stopPointLocation == null) {
                continue
            }
            if (plugin.isPlayerWithinStopRadius(stop, radiusSq)) {
                return stop
            }
        }
        return null
    }

    private fun isTerminalDemand(service: LineService, stopIndex: Int, stops: List<String>): Boolean =
        stopIndex >= stops.size - 1 && !service.isLoopLine

    private fun materializeTrain(
        service: LineService,
        virtualTrain: VirtualTrain,
        targetStopId: String,
        targetIndex: Int,
        stops: List<String>,
        currentTick: Long,
    ): Boolean {
        val plugin = service.plugin
        if (targetIndex >= stops.size - 1 && !service.isLoopLine) {
            plugin.logger.fine("[LocalDispatch] Skipping terminal stop $targetStopId")
            return false
        }
        val fromStopIndex = virtualTrain.currentStopIndex
        val train = service.spawnTrainForVirtual(
            currentTick,
            fromStopIndex,
            fromStopIndex,
            0.0,
            virtualTrain.id,
            targetStopId,
        )
        if (train != null) {
            pool?.markMaterialized(virtualTrain.id)
            return true
        }
        plugin.logger.warning("[LocalDispatch] Failed to spawn train at $targetStopId")
        return false
    }

    fun returnTrainToPool(train: TrainInstance, currentTick: Long) {
        val currentPool = pool ?: return
        val virtualTrainId = train.virtualTrainId ?: return
        val line = train.service.line ?: return
        val state = train.getVirtualizationState(currentTick)
        currentPool.returnToVirtual(
            virtualTrainId,
            state.currentIndex,
            state.targetIndex,
            state.progress,
            state.isWaiting,
            currentTick,
            line.orderedStopIds,
        )
    }

    fun releaseMaterializedTrain(virtualTrainId: UUID?) {
        if (virtualTrainId != null) {
            pool?.releaseMaterialized(virtualTrainId)
        }
    }

    fun isMaterializedTrain(virtualTrainId: UUID?): Boolean =
        virtualTrainId != null && pool?.isMaterialized(virtualTrainId) == true

    override fun refreshTopology(service: LineService, newStopIds: List<String>, currentTick: Long) {
        pool?.refreshTopology(newStopIds, service.plugin.travelTimeEstimator, currentTick)
    }
}
